### PR TITLE
Feature/rpm module

### PR DESCRIPTION
**Pull request by rebeccaskinner**
Refactored for testability and verified on centos; ready for another look.

### CODE REVIEW

File: resource/packages/apt/preparer.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+// Preparer for APT Package
+//
+// APT Package manages system packages
+type Preparer struct {
+	// Name of the system package to be managed.
+	Name string `hcl:"name" required:"true" `
+
+	// State defines desired system package state.
+	State State `hcl:"state" valid_values:"installed,absent" default:"installed"`
```
Comment (ryane) [inline resource/packages/apt/preparer.go]: should this be "present,absent"?

File: resource/packages/apt/apt.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+	Name  string
+	State State
+}
+
+// State type for Package
+type State string
+
+const (
+	// StatePresent indicates the package should be present
+	StatePresent State = "present"
```
Comment (ryane) [inline resource/packages/apt/apt.go]: seems like we are using "present" in some places but "installed" in others

File: resource/packages/rpm/packageManager.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
+// Copyright © 2016 Asteris, LLC
```
Comment (ryane) [inline resource/packages/rpm/packageManager.go]: package_manager.go?

File: resource/packages/rpm/packageManager_test.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
+// Copyright © 2016 Asteris, LLC
```
Comment (ryane) [inline resource/packages/rpm/packageManager_test.go]: package_manager_test.go?

File: resource/packages/rpm/packageManager.go [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+	"github.com/pkg/errors"
+)
+
+// PackageVersion is a type alias for a string, since various packages may use
+// different naming conventions
+type PackageVersion string
+
+// PackageManager describes an interface for managing packages and helps make
+// `Check` and `Apply` testable.
+type PackageManager interface {
```
Comment (ryane) [inline resource/packages/rpm/packageManager.go]: some of the interfaces and functions in this file looks like they could be applicable to the apt resource also

File: resource/packages/rpm/preparer.go [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+// Preparer for RPM Package
+//
+// RPM Package manages system packages
+type Preparer struct {
+	// Name of the system package to be managed.
+	Name string `hcl:"name" required:"true" `
+
+	// State defines desired system package state.
+	State State `hcl:"state" valid_values:"present,absent" default:"installed"`
```
Comment (ryane) [inline resource/packages/rpm/preparer.go]: default:"present"

File: resource/packages/rpm/rpm.go [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+	var err error
+	p.Status = resource.NewStatus()
+	if p.State == p.PackageState() {
+		return p, nil
+	}
+
+	var results string
+	if p.State == StatePresent {
+		results, err = p.PkgMgr.InstallPackage(p.Name)
+		p.Status.AddMessage("installed " + p.Name)
```
Comment (stevendborrelli) [inline resource/packages/rpm/rpm.go]: I think we should also have `addDifference`  here. Right now check and apply runs look different, with `apply` not showing changes:

```
# converge plan --local samples/rpm.hcl 
root/rpm.package.mc:
 Messages:
 Has Changes: yes
 Changes:
  mc gcc git emacs: "absent" => "present"

Summary: 0 errors, 1 changes

# converge apply  --local samples/rpm.hcl 

root/rpm.package.mc:
 Messages:
 Has Changes: yes
 Changes: No changes

Summary: 0 errors, 1 changes
```

File: resource/packages/rpm/packageManager.go [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+// SysCaller allows us to mock exec.Command
+type SysCaller interface {
+	Run(string) ([]byte, error)
+}
+
+// ExecCaller is a dummy struct to handle wrapping exec.Command in the SysCaller
+// interface.
+type ExecCaller struct{}
+
+// Run executs `cmd` as a /bin/sh script and returns the output and error
```
Comment (stevendborrelli) [inline resource/packages/rpm/packageManager.go]: executes

File: resource/packages/rpm/packageManager.go (hunk 5 above)
Comment (rebeccaskinner) [inline resource/packages/rpm/packageManager.go]: yeah, I planned on factoring it out once I tackle the apt module.

File: resource/packages/rpm/rpm.go [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+const (
+	// StatePresent indicates the package should be present
+	StatePresent State = "present"
+
+	// StateAbsent indicates the package should be absent
+	StateAbsent State = "absent"
+)
+
+// Check if the package has to be 'installed', or 'absent'
```
Comment (ryane) [inline resource/packages/rpm/rpm.go]: present

File: resource/packages/rpm/rpm.go (hunk 7 above)
Comment (ryane) [inline resource/packages/rpm/rpm.go]: shouldn't you see "installed mc gcc git emacs" in the Messages: output?

Comment (stevendborrelli) [inline resource/packages/rpm/rpm.go]: This is a bug in the engine itself, so probably isn't a blocker here.

**Pull request merged**